test/unitTest/plan/executrix/configurator.txt
Description: Test plan for configurator.js

--------------------------------------------------------------------------------
testName: setConfigurationSetting_validData
input configurationNamespace: string (The path in the configuration JSON object where the configuration setting should be set)
input configurationName: string (The key of the configuration setting)
input configurationValue: string|integer|boolean|double (The value of the configuration setting)
output: void

testName: setConfigurationSetting_inValidConfigurationNamespaceString
input configurationNamespace: string (dfxg24346dfg)
input configurationName: string (The key of the configuration setting)
input configurationValue: string|integer|boolean|double (The value of the configuration setting)
output: ?

testName: setConfigurationSetting_inValidConfigurationnameString
input configurationNamespace: string (The path in the configuration JSON object where the configuration setting should be set)
input configurationName: string (dfxg24346dfg)
input configurationValue: string|integer|boolean|double (The value of the configuration setting)
output: ?

testName: setConfigurationSetting_inValidConfigurationValueString
input configurationNamespace: string (The path in the configuration JSON object where the configuration setting should be set)
input configurationName: string (The key of the configuration setting)
input configurationValue: string (dfxg24346dfg)
output: ?

testName: setConfigurationSetting_inValidConfigurationNamespaceInteger
input configurationNamespace: integer (123)
input configurationName: string (The key of the configuration setting)
input configurationValue: string|integer|boolean|double (The value of the configuration setting)
output: ?

testName: setConfigurationSetting_inValidConfigurationNamespaceBoolean
input configurationNamespace: boolean (false)
input configurationName: string (The key of the configuration setting)
input configurationValue: string|integer|boolean|double (The value of the configuration setting)
output: ?

testName: setConfigurationSetting_inValidConfigurationNamespaceUndefined
input configurationNamespace: undefined
input configurationName: string (The key of the configuration setting)
input configurationValue: string|integer|boolean|double (The value of the configuration setting)
output: ?

testName: setConfigurationSetting_inValidConfigurationNamespaceNaN
input configurationNamespace: NaN
input configurationName: string (The key of the configuration setting)
input configurationValue: string|integer|boolean|double (The value of the configuration setting)
output: ?

testName: setConfigurationSetting_inValidConfigurationNameInteger
input configurationNamespace: string (The path in the configuration JSON object where the configuration setting should be set)
input configurationName: integer (123)
input configurationValue: string|integer|boolean|double (The value of the configuration setting)
output: ?

testName: setConfigurationSetting_inValidConfigurationNameBoolean
input configurationNamespace: string (The path in the configuration JSON object where the configuration setting should be set)
input configurationName: boolean (false)
input configurationValue: string|integer|boolean|double (The value of the configuration setting)
output: ?

testName: setConfigurationSetting_inValidConfigurationNameUndefined
input configurationNamespace: string (The path in the configuration JSON object where the configuration setting should be set)
input configurationName: undefined
input configurationValue: string|integer|boolean|double (The value of the configuration setting)
output: ?

testName: setConfigurationSetting_inValidConfigurationNameNaN
input configurationNamespace: string (The path in the configuration JSON object where the configuration setting should be set)
input configurationName: NaN
input configurationValue: string|integer|boolean|double (The value of the configuration setting)
output: ?

testName: setConfigurationSetting_inValidConfigurationValueInteger
input configurationNamespace: string (The path in the configuration JSON object where the configuration setting should be set)
input configurationName: string (The key of the configuration setting)
input configurationValue: integer (123)
output: ?

testName: setConfigurationSetting_inValidConfigurationValueString
input configurationNamespace: string (The path in the configuration JSON object where the configuration setting should be set)
input configurationName: string (The key of the configuration setting)
input configurationValue: boolean (false)
output: ?

testName: setConfigurationSetting_inValidConfigurationValueUndefined
input configurationNamespace: string (The path in the configuration JSON object where the configuration setting should be set)
input configurationName: string (The key of the configuration setting)
input configurationValue: undefined
output: ?

testName: setConfigurationSetting_inValidConfigurationValueNaN
input configurationNamespace: string (The path in the configuration JSON object where the configuration setting should be set)
input configurationName: string (The key of the configuration setting)
input configurationValue: NaN
output: ?

testName: setConfigurationSetting_inValidAllUndefined
input configurationNamespace: undefined
input configurationName: undefined
input configurationValue: undefined
output: ?

testName: setConfigurationSetting_inValidAllNaN
input configurationNamespace: NaN
input configurationName: NaN
input configurationValue: NaN
output: ?
--------------------------------------------------------------------------------
testName: setPluginConfigurationSetting_validData
input dataStructure: object (The input data structure upon which the configuration setting should be set and then returned)
input configurationNamespace: string (The path in the configuration JSON object where the configuration setting should be set)
input configurationName: string (The key of the configuration setting)
input configurationValue: string|integer|boolean|double (The value of the configuration setting)
output: object (The modified input object with the new configuration value added in the appropriate location)

testName: setPluginConfigurationSetting_inValidDataStructureString
input dataStructure: string (dfxg24346dfg)
input configurationNamespace: string (The path in the configuration JSON object where the configuration setting should be set)
input configurationName: string (The key of the configuration setting)
input configurationValue: string|integer|boolean|double (The value of the configuration setting)
output: ?

testName: setPluginConfigurationSetting_inValidConfigurationNamespaceString
input dataStructure: object (The input data structure upon which the configuration setting should be set and then returned)
input configurationNamespace: string (dfxg24346dfg)
input configurationName: string (The key of the configuration setting)
input configurationValue: string|integer|boolean|double (The value of the configuration setting)
output: ?

testName: setPluginConfigurationSetting_inValidConfigurationnameString
input dataStructure: object (The input data structure upon which the configuration setting should be set and then returned)
input configurationNamespace: string (The path in the configuration JSON object where the configuration setting should be set)
input configurationName: string (dfxg24346dfg)
input configurationValue: string|integer|boolean|double (The value of the configuration setting)
output: ?

testName: setPluginConfigurationSetting_inValidConfigurationValueString
input dataStructure: object (The input data structure upon which the configuration setting should be set and then returned)
input configurationNamespace: string (The path in the configuration JSON object where the configuration setting should be set)
input configurationName: string (The key of the configuration setting)
input configurationValue: string|integer|boolean|double (dfxg24346dfg)
output: ?

testName: setPluginConfigurationSetting_inValidDataStructureInteger
input dataStructure: integer (123)
input configurationNamespace: string (The path in the configuration JSON object where the configuration setting should be set)
input configurationName: string (The key of the configuration setting)
input configurationValue: string|integer|boolean|double (The value of the configuration setting)
output: ?

testName: setPluginConfigurationSetting_inValidDataStructureBoolean
input dataStructure: boolean (false)
input configurationNamespace: string (The path in the configuration JSON object where the configuration setting should be set)
input configurationName: string (The key of the configuration setting)
input configurationValue: string|integer|boolean|double (The value of the configuration setting)
output: ?

testName: setPluginConfigurationSetting_inValidDataStructureUndefined
input dataStructure: undefined
input configurationNamespace: string (The path in the configuration JSON object where the configuration setting should be set)
input configurationName: string (The key of the configuration setting)
input configurationValue: string|integer|boolean|double (The value of the configuration setting)
output: ?

testName: setPluginConfigurationSetting_inValidDataStructureNaN
input dataStructure: NaN
input configurationNamespace: string (The path in the configuration JSON object where the configuration setting should be set)
input configurationName: string (The key of the configuration setting)
input configurationValue: string|integer|boolean|double (The value of the configuration setting)
output: ?

testName: setPluginConfigurationSetting_inValidConfigurationNamespaceInteger
input dataStructure: object (The input data structure upon which the configuration setting should be set and then returned)
input configurationNamespace: integer (123)
input configurationName: string (The key of the configuration setting)
input configurationValue: string|integer|boolean|double (The value of the configuration setting)
output: ?

testName: setPluginConfigurationSetting_inValidConfigurationNamespaceBoolean
input dataStructure: object (The input data structure upon which the configuration setting should be set and then returned)
input configurationNamespace: boolean (false)
input configurationName: string (The key of the configuration setting)
input configurationValue: string|integer|boolean|double (The value of the configuration setting)
output: ?

testName: setPluginConfigurationSetting_inValidConfigurationNamespaceUndefined
input dataStructure: object (The input data structure upon which the configuration setting should be set and then returned)
input configurationNamespace: undefined
input configurationName: string (The key of the configuration setting)
input configurationValue: string|integer|boolean|double (The value of the configuration setting)
output: ?

testName: setPluginConfigurationSetting_inValidConfigurationNamespaceNaN
input dataStructure: object (The input data structure upon which the configuration setting should be set and then returned)
input configurationNamespace: NaN
input configurationName: string (The key of the configuration setting)
input configurationValue: string|integer|boolean|double (The value of the configuration setting)
output: ?

testName: setPluginConfigurationSetting_inValidConfigurationNameInteger
input dataStructure: object (The input data structure upon which the configuration setting should be set and then returned)
input configurationNamespace: string (The path in the configuration JSON object where the configuration setting should be set)
input configurationName: integer (123)
input configurationValue: string|integer|boolean|double (The value of the configuration setting)
output: ?

testName: setPluginConfigurationSetting_inValidConfigurationNameBoolean
input dataStructure: object (The input data structure upon which the configuration setting should be set and then returned)
input configurationNamespace: string (The path in the configuration JSON object where the configuration setting should be set)
input configurationName: boolean (false)
input configurationValue: string|integer|boolean|double (The value of the configuration setting)
output: ?

testName: setPluginConfigurationSetting_inValidConfigurationNameUndefined
input dataStructure: object (The input data structure upon which the configuration setting should be set and then returned)
input configurationNamespace: string (The path in the configuration JSON object where the configuration setting should be set)
input configurationName: undefined
input configurationValue: string|integer|boolean|double (The value of the configuration setting)
output: ?

testName: setPluginConfigurationSetting_inValidConfigurationNameNaN
input dataStructure: object (The input data structure upon which the configuration setting should be set and then returned)
input configurationNamespace: string (The path in the configuration JSON object where the configuration setting should be set)
input configurationName: NaN
input configurationValue: string|integer|boolean|double (The value of the configuration setting)
output: ?

testName: setPluginConfigurationSetting_inValidConfigurationValueInteger
input dataStructure: object (The input data structure upon which the configuration setting should be set and then returned)
input configurationNamespace: string (The path in the configuration JSON object where the configuration setting should be set)
input configurationName: string (The key of the configuration setting)
input configurationValue: integer (123)
output: ?

testName: setPluginConfigurationSetting_inValidConfigurationValueBoolean
input dataStructure: object (The input data structure upon which the configuration setting should be set and then returned)
input configurationNamespace: string (The path in the configuration JSON object where the configuration setting should be set)
input configurationName: string (The key of the configuration setting)
input configurationValue: boolean (false)
output: ?

testName: setPluginConfigurationSetting_inValidConfigurationValueUndefined
input dataStructure: object (The input data structure upon which the configuration setting should be set and then returned)
input configurationNamespace: string (The path in the configuration JSON object where the configuration setting should be set)
input configurationName: string (The key of the configuration setting)
input configurationValue: undefined
output: ?

testName: setPluginConfigurationSetting_inValidConfigurationValueNaN
input dataStructure: object (The input data structure upon which the configuration setting should be set and then returned)
input configurationNamespace: string (The path in the configuration JSON object where the configuration setting should be set)
input configurationName: string (The key of the configuration setting)
input configurationValue: NaN
output: ?

testName: setPluginConfigurationSetting_inValidAllUndefined
input dataStructure: undefined
input configurationNamespace: undefined
input configurationName: undefined
input configurationValue: undefined
output: ?

testName: setPluginConfigurationSetting_inValidAllNaN
input dataStructure: NaN
input configurationNamespace: NaN
input configurationName: NaN
input configurationValue: NaN
output: ?
--------------------------------------------------------------------------------
testName: getConfigurationSetting_validData
input configurationNamespace: string (The path in the configuration JSON object where the configuration setting should be found)
input configurationName: string (The key of the configuration setting)
output: string|integer|boolean|double (The value of whatever was stored in the D[configuration])

testName: getConfigurationSetting_inValidConfigurationNamespaceString
input configurationNamespace: string (dfxg24346dfg)
input configurationName: string (The key of the configuration setting)
output: ?

testName: getConfigurationSetting_inValidConfigurationNameString
input configurationNamespace: string (The path in the configuration JSON object where the configuration setting should be found)
input configurationName: string (dfxg24346dfg)
output: ?

testName: getConfigurationSetting_inValidConfigurationNamespaceInteger
input configurationNamespace: integer (123)
input configurationName: string (The key of the configuration setting)
output: ?

testName: getConfigurationSetting_inValidConfigurationNamespaceBoolean
input configurationNamespace: boolean (false)
input configurationName: string (The key of the configuration setting)
output: ?

testName: getConfigurationSetting_inValidConfigurationNameInteger
input configurationNamespace: string (The path in the configuration JSON object where the configuration setting should be found)
input configurationName: integer (123)
output: ?

testName: getConfigurationSetting_inValidConfigurationNameBoolean
input configurationNamespace: string (The path in the configuration JSON object where the configuration setting should be found)
input configurationName: boolean (false)
output: ?

testName: getConfigurationSetting_inValidConfigurationNamespaceUndefined
input configurationNamespace: undefined
input configurationName: string (The key of the configuration setting)
output: ?

testName: getConfigurationSetting_inValidConfigurationNamespaceNaN
input configurationNamespace: NaN
input configurationName: string (The key of the configuration setting)
output: ?

testName: getConfigurationSetting_inValidConfigurationNameUndefiend
input configurationNamespace: string (The path in the configuration JSON object where the configuration setting should be found)
input configurationName: undefined
output: ?

testName: getConfigurationSetting_inValidConfigurationNameNaN
input configurationNamespace: string (The path in the configuration JSON object where the configuration setting should be found)
input configurationName: NaN
output: ?
--------------------------------------------------------------------------------
testName: processConfigurationNameRules_validData
input fullyQualifiedName: string (The fully qualified name with the namespace included)
output: string (The name of the configuration setting without the namespace)

testName: processConfigurationNameRules_inValidProcessConfigurationNameRulesString
input fullyQualifiedName: string (dfxg24346dfg)
output: ?

testName: processConfigurationNameRules_inValidProcessConfigurationNameRulesInteger
input fullyQualifiedName: integer (123)
output: ?

testName: processConfigurationNameRules_inValidProcessConfigurationNameRulesBoolean
input fullyQualifiedName: boolean (false)
output: ?

testName: processConfigurationNameRules_inValidProcessConfigurationNameRulesUndefined
input fullyQualifiedName: undefined
output: ?

testName: processConfigurationNameRules_inValidProcessConfigurationNameRulesNaN
input fullyQualifiedName: NaN
output: ?
--------------------------------------------------------------------------------
testName: processConfigurationNamespaceRules_validData
input fullyQualifiedName: string (The fully qualified name with the namespace included)
output: string (The namespace of the configuration setting, without the configuration name)

testName: processConfigurationNamespaceRules_inValidProcessConfigurationNameRulesString
input fullyQualifiedName: string (dfxg24346dfg)
output: ?

testName: processConfigurationNamespaceRules_inValidProcessConfigurationNameRulesInteger
input fullyQualifiedName: integer (123)
output: ?

testName: processConfigurationNamespaceRules_inValidProcessConfigurationNameRulesBoolean
input fullyQualifiedName: boolean (false)
output: ?

testName: processConfigurationNamespaceRules_inValidProcessConfigurationNameRulesUndefined
input fullyQualifiedName: undefined
output: ?

testName: processConfigurationNamespaceRules_inValidProcessConfigurationNameRulesNaN
input fullyQualifiedName: NaN
output: ?
--------------------------------------------------------------------------------
testName: processConfigurationValueRules_validData
input name: string (The name of the configuration variable, without the namespace)
input value: string (The value of the configuration variable)
output: string|boolean|integer|float|object (A value that is appropriately processed)

testName: processConfigurationValueRules_inValidNameString
input name: string (dfxg24346dfg)
input value: string (The value of the configuration variable)
output: ?

testName: processConfigurationValueRules_inValidValueString
input name: string (The name of the configuration variable, without the namespace)
input value: string (dfxg24346dfg)
output: ?

testName: processConfigurationValueRules_inValidNameInteger
input name: integer (123)

CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

### [0.2.11](https://github.com/SethEden/haystacks-async/compare/v0.2.10...v0.2.11) (2023-04-28)

### [0.2.10](https://github.com/SethEden/haystacks-async/compare/v0.2.9...v0.2.10) (2023-04-26)

### [0.2.9](https://github.com/SethEden/haystacks-async/compare/v0.2.8...v0.2.9) (2023-04-17)

### [0.2.8](https://github.com/SethEden/haystacks-async/compare/v0.2.7...v0.2.8) (2023-04-06)

### [0.2.7](https://github.com/SethEden/haystacks-async/compare/v0.2.6...v0.2.7) (2023-03-29)

### [0.2.6](https://github.com/SethEden/haystacks-async/compare/v0.2.5...v0.2.6) (2023-03-28)

### [0.2.5](https://github.com/SethEden/haystacks-async/compare/v0.2.4...v0.2.5) (2023-03-17)

### [0.2.4](https://github.com/SethEden/haystacks-async/compare/v0.2.3...v0.2.4) (2023-03-15)

### [0.2.3](https://github.com/SethEden/haystacks-async/compare/v0.2.2...v0.2.3) (2023-03-14)

### [0.2.2](https://github.com/SethEden/haystacks-async/compare/v0.2.0...v0.2.2) (2023-03-14)

### [0.1.12](https://github.com/SethEden/haystacks-async/compare/v0.1.11...v0.1.12) (2023-02-18)

### [0.1.11](https://github.com/SethEden/haystacks-async/compare/v0.1.10...v0.1.11) (2023-02-15)

### [0.1.12](https://github.com/SethEden/haystacks-async/compare/v0.1.10...v0.1.12) (2023-02-15)

### [0.1.11](https://github.com/SethEden/haystacks-async/compare/v0.1.10...v0.1.11) (2023-02-12)

### [0.1.10](https://github.com/SethEden/haystacks-async/compare/v0.1.9...v0.1.10) (2023-02-12)

### [0.1.9](https://github.com/SethEden/haystacks-async/compare/v0.1.8...v0.1.9) (2023-02-09)

### [0.1.8](https://github.com/SethEden/haystacks-async/compare/v0.1.7...v0.1.8) (2023-01-11)

### [0.1.7](https://github.com/SethEden/haystacks-async/compare/v0.1.6...v0.1.7) (2023-01-10)

### [0.1.6](https://github.com/SethEden/haystacks-async/compare/v0.1.5...v0.1.6) (2022-12-29)

### [0.1.5](https://github.com/SethEden/haystacks-async/compare/v0.1.4...v0.1.5) (2022-10-26)

### [0.1.4](https://github.com/SethEden/haystacks-async/compare/v0.1.3...v0.1.4) (2022-09-22)

### [0.1.3](https://github.com/SethEden/haystacks-async/compare/v0.1.2...v0.1.3) (2022-09-20)

### [0.1.2](https://github.com/SethEden/haystacks-async/compare/v0.1.1...v0.1.2) (2022-09-01)

### 0.1.1 (2022-06-24)

### 0.1.1 (2022-06-24)

### [0.0.17](https://github.com/SethEden/haystacks/compare/v0.0.16...v0.0.17) (2022-06-22)

### [0.0.16](https://github.com/SethEden/haystacks/compare/v0.0.15...v0.0.16) (2022-06-15)

### [0.0.15](https://github.com/SethEden/haystacks/compare/v0.0.14...v0.0.15) (2022-06-15)

### [0.0.14](https://github.com/SethEden/haystacks/compare/v0.0.13...v0.0.14) (2022-06-10)

### [0.0.13](https://github.com/SethEden/haystacks/compare/v0.0.12...v0.0.13) (2022-06-08)

### [0.0.12](https://github.com/SethEden/haystacks/compare/v0.0.11...v0.0.12) (2022-05-26)

### [0.0.11](https://github.com/SethEden/haystacks/compare/v0.0.10...v0.0.11) (2022-05-12)

### [0.0.10](https://github.com/SethEden/haystacks/compare/v0.0.9...v0.0.10) (2022-05-10)

### [0.0.9](https://github.com/SethEden/haystacks/compare/v0.0.8...v0.0.9) (2022-04-28)

### [0.0.8](https://github.com/SethEden/haystacks/compare/v0.0.7...v0.0.8) (2022-04-13)

### [0.0.7](https://github.com/SethEden/haystacks/compare/v0.0.3...v0.0.7) (2022-04-10)

### [0.0.5](https://github.com/SethEden/haystacks/compare/v0.0.3...v0.0.5) (2021-10-28)

### [0.0.3](https://github.com/SethEden/haystacks/compare/v0.0.2...v0.0.3) (2021-10-27)

### 0.0.2 (2021-10-26)


src/commandsBlob/commands/configuration.js
/**
 * @file configuration.js
 * @module configuration
 * @description Contains all of the configuration system commands.
 * @requires module:dataBroker
 * @requires module:ruleBroker
 * @requires module:themeBroker
 * @requires module:configurator
 * @requires module:loggers
 * @requires module:data
 * @requires {@link https://www.npmjs.com/package/@haystacks/constants|@haystacks/constants}
 * @requires {@link https://www.npmjs.com/package/path|path}
 * @author Seth Hollingsead
 * @date 2022/02/04
 * @copyright Copyright © 2022-… by Seth Hollingsead. All rights reserved
 */

// Internal imports
import dataBroker from '../../brokers/dataBroker.js';
import ruleBroker from '../../brokers/ruleBroker.js';
import themeBroker from '../../brokers/themeBroker.js';
import configurator from '../../executrix/configurator.js';
import loggers from '../../executrix/loggers.js';
import D from '../../structures/data.js';
// External imports
import hayConst from '@haystacks/constants';
import path from 'path';

const {bas, biz, cfg, gen, msg, sys, wrd} = hayConst;
const baseFileName = path.basename(import.meta.url, path.extname(import.meta.url));
// framework.commandsBlob.commands.configuration.
const namespacePrefix = wrd.cframework + bas.cDot + sys.ccommandsBlob + bas.cDot + wrd.ccommands + bas.cDot + baseFileName + bas.cDot;

/**
 * @function saveConfiguration
 * @description Saves out all of the configuration data to a JSON file so custom user settings can be persisted between sessions.
 * @param {string} inputData Not used for this command.
 * @param {string} inputMetaData Not used for this command.
 * @return {array<boolean,string|integer|boolean|object|array>} An array with a boolean True or False value to
 * indicate if the application should exit or not exit, followed by the command output.
 * @author Seth Hollingsead
 * @date 2022/03/11
 */
async function saveConfiguration(inputData, inputMetaData) {
  let functionName = saveConfiguration.name;
  loggers.consoleLog(namespacePrefix + functionName, msg.cBEGIN_Function);
  loggers.consoleLog(namespacePrefix + functionName, msg.cinputDataIs + JSON.stringify(inputData));
  loggers.consoleLog(namespacePrefix + functionName, msg.cinputMetaDataIs + inputMetaData);
  let returnData = [true, {}];
  returnData[1] = await dataBroker.writeJsonDataToFile(await configurator.getConfigurationSetting(wrd.csystem, cfg.cappConfigPath) + wrd.cconfig + gen.cDotjson, JSON.stringify(D));
  loggers.consoleLog(namespacePrefix + functionName, msg.creturnDataIs + JSON.stringify(returnData));
  loggers.consoleLog(namespacePrefix + functionName, msg.cEND_Function);
  return returnData;
}

/**
 * @function changeConfigurationSetting
 * @description Persists a change to a user specified configuration setting.
 * This enables the user to enable or disable console log flags to use for run-time debugging.
 * This is useful also for automating changes to how and what is logged during execution.
 * Further it is useful for when a test fails, it can be automatically re-run with a more detailed log.
 * @param {array<string>} inputData An array that contains
 * the fully qualified path to the configuration setting that should be changed and
 * the data that should be assigned to it.
 * inputData[0] = changeConfigurationSetting
 * inputData[1] = fully.Qualified.Configuration.Path
 * inputData[2] = value to assign to the configuration setting property
 * @param {string} inputMetaData Not used for this command.
 * @return {array<boolean,string|integer|boolean|object|array>} An array with a boolean True or False value to
 * indicate if the application should exit or not exit, followed by the command output.
 * @author Seth Hollingsead
 * @date 2022/05/11
 * @NOTE Test String 1: changeConfigurationSetting configuration.debugSetting.commandsBlob.commands.system true
 * Test String 2: changeConfigurationSetting debugSetting.commandsBlob.commands.system true
 */
async function changeConfigurationSetting(inputData, inputMetaData) {
  let functionName = changeConfigurationSetting.name;
  loggers.consoleLog(namespacePrefix + functionName, msg.cBEGIN_Function);
  loggers.consoleLog(namespacePrefix + functionName, msg.cinputDataIs + JSON.stringify(inputData));
  loggers.consoleLog(namespacePrefix + functionName, msg.cinputMetaDataIs + inputMetaData);
  let returnData = [true, {}];
  let errorMessage = '';
  if (inputData && inputData.length === 3) {
    let dataPath = inputData[1];
    dataPath = await ruleBroker.processRules([dataPath, ''], [biz.cgetWordsArrayFromString]);
    // dataPath is:
    loggers.consoleLog(namespacePrefix + functionName, msg.cdataPathIs + JSON.stringify(dataPath));
    let newValue = inputData[2];
    // newValue is:
    loggers.consoleLog(namespacePrefix + functionName, msg.cnewValueIs + JSON.stringify(newValue));
    if (dataPath[0] === wrd.cconfiguration) {
      dataPath.shift(wrd.cconfiguration);
    }
    let configurationName = dataPath.pop();
    // dataPath is:
    loggers.consoleLog(namespacePrefix + functionName, msg.cdataPathIs + JSON.stringify(dataPath));
    dataPath = dataPath.join(bas.cDot);
    newValue = await ruleBroker.processRules([newValue, ''], [biz.cstringToDataType]);
    await configurator.setConfigurationSetting(dataPath, configurationName, newValue);
    returnData[1] = true;
  } else {
    // ERROR: Invalid entry, please enter a valid configuration namespace to change,
    // and a value to assign to the configuration setting.
    errorMessage = msg.cchangeConfigurationSettingMessage01 + msg.cchangeConfigurationSettingMessage02;
    console.log(errorMessage);
    returnData[1] = errorMessage;
    // EXAMPLE: changeConfigurationSetting debugSetting.businessRules.rules.arrayParsing.commandArrayParsing.solveLehmerCode true
    console.log(msg.cchangeConfigurationSettingMessage03);
  }
  loggers.consoleLog(namespacePrefix + functionName, msg.creturnDataIs + JSON.stringify(returnData));
  loggers.consoleLog(namespacePrefix + functionName, msg.cEND_Function);
  return returnData;
}

/**
 * @function listConfigurationThemes
 * @description Lists all of the debug configuration themes currently installed in the resources/themes folder.
 * @param {string} inputData Not used for this command.
 * @param {string} inputMetaData Not used for this command.
 * @return {array<boolean,string|integer|boolean|object|array>} An array with a boolean True or False value to
 * indicate if the application should exit or not exit, followed by the command output.
 * @author Seth Hollingsead
 * @date 2022/06/10
 */
async function listConfigurationThemes(inputData, inputMetaData) {
  let functionName = listConfigurationThemes.name;
  loggers.consoleLog(namespacePrefix + functionName, msg.cBEGIN_Function);
  loggers.consoleLog(namespacePrefix + functionName, msg.cinputDataIs + JSON.stringify(inputData));
  loggers.consoleLog(namespacePrefix + functionName, msg.cinputMetaDataIs + inputMetaData);
  let returnData = [true, {}];
  let frameworkThemesPath = await configurator.getConfigurationSetting(wrd.csystem, cfg.cframeworkThemesPath);
  frameworkThemesPath = path.resolve(frameworkThemesPath);
  // frameworkThemesPath is:
  loggers.consoleLog(namespacePrefix + functionName, msg.cframeworkThemesPathIs + frameworkThemesPath);
  let themesList = await themeBroker.getNamedThemesFromRootPath(frameworkThemesPath);
  // themesList is:
  console.log(msg.cthemesListIs + JSON.stringify(themesList));
  returnData[1] = themesList;
  loggers.consoleLog(namespacePrefix + functionName, msg.creturnDataIs + JSON.stringify(returnData));
  loggers.consoleLog(namespacePrefix + functionName, msg.cEND_Function);
  return returnData;
}

/**
 * @function changeDebugConfigurationTheme
 * @description Swaps out all of the debug settings with a different theme.
 * @param {string} inputData An array that contains the name of the theme the user would like to switch to.
 * inputData[0] = changeDebugConfigurationTheme
 * inputData[1] = The name of the theme that the user would like to switch to.
 * @param {string} inputMetaData Not used for this command.
 * @return {array<boolean,string|integer|boolean|object|array>} An array with a boolean True or False value to
 * indicate if the application should exit or not exit, followed by the command output.
 * @author Seth Hollingsead
 * @date 2022/06/13
 */
async function changeDebugConfigurationTheme(inputData, inputMetaData) {
  let functionName = changeDebugConfigurationTheme.name;
  loggers.consoleLog(namespacePrefix + functionName, msg.cBEGIN_Function);
  loggers.consoleLog(namespacePrefix + functionName, msg.cinputDataIs + JSON.stringify(inputData));
  loggers.consoleLog(namespacePrefix + functionName, msg.cinputMetaDataIs + inputMetaData);
  let returnData = [true, {}];
  let errorMessage = '';
  if (inputData && inputData.length === 2) {
    let desiredThemeName = inputData[1];
    // desiredThemeName is:
    loggers.consoleLog(namespacePrefix + functionName, msg.cdesiredThemeNameIs + desiredThemeName);
    let frameworkThemesPath = await configurator.getConfigurationSetting(wrd.csystem, cfg.cframeworkThemesPath);
    frameworkThemesPath = path.resolve(frameworkThemesPath);
    // frameworkThemesPath is:
    loggers.consoleLog(namespacePrefix + functionName, msg.cframeworkThemesPathIs + frameworkThemesPath);
    let namedThemePath = themeBroker.getNamedThemePathFromRootPath(desiredThemeName, frameworkThemesPath);
    if (namedThemePath !== false) {
      // namedThemePath is verified:
      loggers.consoleLog(namespacePrefix + functionName, msg.cnamedThemePathIsVerified + namedThemePath);
      await configurator.setConfigurationSetting(wrd.csystem, sys.cthemeConfigPath, namedThemePath);
      let loadedThemeData = await themeBroker.loadTheme(namedThemePath);
      // loadedThemeData is:
      loggers.consoleLog(namespacePrefix + functionName, msg.cloadedThemeDataIs + JSON.stringify(loadedThemeData));
      let themeLoadedSuccessfully = await themeBroker.applyTheme(loadedThemeData);
      returnData[1] = themeLoadedSuccessfully;
      if (themeLoadedSuccessfully === false) {
        // ERROR: There was an error applying the selected theme to the active debug settings configuration.
        errorMessage = msg.cchangeDebugConfigurationThemeMessage01;
        console.log(errorMessage);
        returnData[1] = errorMessage;
      } // End-if (themeLoadedSuccessfully === false)
    } else {
      // ERROR: The specified theme name was not found in the current list of supported themes.
      errorMessage = msg.cchangeDebugConfigurationThemeMessage02;
      console.log(errorMessage);
      returnData[1] = errorMessage;
      // You can find the available themes at the following path location:
      console.log(msg.cchangeDebugConfigurationThemeMessage03 +
        await configurator.getConfigurationSetting(wrd.csystem, cfg.cframeworkThemesPath));
    }
  } else {
    // ERROR: Invalid entry, please enter a theme name you would like the debug settings to switch to when logging debug statements.
    errorMessage = msg.cchangeDebugConfigurationThemeMessage04
    console.log(errorMessage);
    returnData[1] = errorMessage;
    // EXAMPLE: changeDebugConfigurationTheme Skywalker
    console.log(msg.cchangeDebugConfigurationThemeMessage05);
  }
  loggers.consoleLog(namespacePrefix + functionName, msg.creturnDataIs + JSON.stringify(returnData));
  loggers.consoleLog(namespacePrefix + functionName, msg.cEND_Function);
  return returnData;
}

export default {
  saveConfiguration,
  changeConfigurationSetting,
  listConfigurationThemes,
  changeDebugConfigurationTheme
};
